rfcs/2022-08-25-12217-handling-discarded-events.md

```diff
@@ -0,0 +1,369 @@
+- Handling of discarded sink batch buffers.
+
+## Pain
+
+Vector has a number of ways in which components may drop events due that are outside of the
+operator's control:
+
+- Transform processing
+- Sink encoding failure
+- Sink partitioning failure
+- Failure to write to disk buffer
+- Incompatible event type
+- Unhandled metric value type
+- Unwired component output
+- Failure to send to next component
+
+This leads to the potential for lost events and even silently dropped events. Users want to be able
+to capture discarded events at various points in the topology in order to forward them to fail-safe
+destinations for diagnosis.
+
+## Proposal
+
+There are three parts to the proposed solution:
+
+1. Add a new standardized output to all components that may discard events.
+2. Update configuration validation to require that all outputs are handled.
+3. Add a new configuration for discarding or rejecting outputs.
+
+### User Experience
+
+#### Add Discarded Event Output
+
+We will introduce a new output to all components that would otherwise discard events, named
+`discards`.
+
+#### Validate Handling All Outputs
+
+Additionally, we will enhance the configuration validation to determine if all outputs are
+handled. This will provide notification of the new discard outputs and ensure users are aware of the
+new feature.
+
+In order to avoid breaking existing configurations, this validation will initially produce a
+deprecation warning, notifying the user that an output is present but unhandled. In a later version,
+this deprecation will be turned into a hard error, preventing the acceptance of a configuration with
+unhandled outputs.
+
+We will also add a new command-line option and environment variable to opt into the stricter
+validation for users that want the additional assurance this provides.
+
+#### Simplify Discarded Output Handling
+
+This enforced handling will create considerable complication for most users, as well as the
+potential of increased overhead and reduced performance if extra components have to be configured
+just to handle the outputs that now need to be connected somewhere. To reduce this overhead, we will
+add a new optional configuration setting to indicate the internal disposition of each output:
+`outputs.NAME.disposition`. This will have two non-default values, `"drop"` to mark all events going
+to that output as delivered and then drop them, and `"reject"` to mark them as having failed.  If
+this setting is not present and a component has an unhandled output, a default behavior of `"drop"`
+will be assumed, which matches the current behavior.
+
+So, to configure the new discard output as discarding events with an error result, users would add the
+following:
+
+```toml
+[transforms.example]
+type = "remap"
+…
+outputs.discards.disposition = "reject"
+```
+
+### Implementation
+
+#### Configuration
+
+The new options will show up in all of the enclosing configuration option structures:
```

Comment:
If we put them on the `Outer` types, they'll be available whether or not the inner component actually discards events. Is that what we want, or should the config only apply to components that discard?

Reply:
I had hoped to avoid having another situation with that like we do for acknowledgements with the `fn can_acknowledge` in the trait. I suppose since some components will not have any output (ex `blackhole` sink), we will have to though. I will update this.